### ISSUE TITLE
Machine generate: doesn't embed JSON children, uses path members

### ISSUE THREAD

**Issue by jonator** [inlang/monorepo#874]
## Problem

https://github.com/osmosis-labs/osmosis-frontend/pull/1622/files/809411475fc2ef0edcbfefb69675af10a657ec85#diff-96148166b124a270d5e3831559a1d4dab4ff671bba2e2072b1e69cb65eb955f6

Looks like machine translate appends AI-generated translations to the end of the file as `errors.fallbackText1: "..."` instead of adding it as a member to the child object like `{ errors: { fallbackText1: "..." } }`. Could cause problems when diffing the files.

## Steps to reproduce

1. Add a JSON child translation in the ref lang
2. Run machine translate
3. Notice key path members

### Additional information (screenshots?)
- Browser:
- Operating system:

**Comment by felixhaeberle** [comment]
@jonator Are you using the CLI here?

**Comment by NilsJacobsen** [comment]
I would say if @jonator is using the editor to male machine translations it is a editor topic. If he uses the CLI then it could be a CLI topic.

**Comment by jonator** [comment]
We use the CLI. WDYM by use the editor, do you have a VSCode extension ? (!!)

**Comment by felixhaeberle** [comment]
@jonator I'll investigate and come back to you. This is a bug and should be fixed.

Yes, we have a VSCode extension :) We're currently reworking the config of it in order to support most of the codebases already using inlang. JSON plugin will probably not get support for the IDE extension **ootb** but you can make it work with a bit of extra work on your side. We have an example of it over here:  https://marketplace.visualstudio.com/items?itemName=inlang.vs-code-extension
I can assist you with setting it up :)

**Comment by jonator** [comment]
OK. I'm out of bandwidth the next few weeks, but great to hear the idea is becoming a reality. 🚀

**Comment by jonator** [comment]
Also I'll be in berlin next 2 weeks starting tomorrow, perhaps we can set it up in person!

**Comment by felixhaeberle** [comment]
Assigning @NilsJacobsen after debugging -> probably the problem is in the json plugin. 
Test with `npx inlang@latest machine translate -f` in the `test-app` example of the `sdk-js`

**Comment by felixhaeberle** [comment]
@samuelstroschein After deprecating `ast.metadata`, how will the machine translate process now give an indicator into the plugin that the message should be nested based on the key? At the moment, machine translate is rather unfortunate as described in this issue, because it does not respect the structure of the json and does not properly nest the output.

See:

<img width="766" alt="Bildschirmfoto 2023-06-22 um 21 33 29" src="https://github.com/inlang/inlang/assets/34959078/a56086d4-4b0c-4c73-ad7f-248d5039a505">

**Comment by samuelstroschein** [comment]
@felixhaeberle I will update the plugin to determine a default for new messages. 

FYI this is unrelated to metadata.

**Comment by felixhaeberle** [comment]
@samuelstroschein Then I can't follow here 😅 Because I thought exactly that was the purpose of metadata to have an understanding of how the message is nested to "unflatten" it out of the AST into the resource at the right position.

TLDR: The messages should be nested correctly into the existing structure and not be inserted "flattened".

**Comment by samuelstroschein** [comment]
@felixhaeberle the state (information) on how keys are nested exists in plugin scope now. 

Metadata broke this because creating new messages, like machine translations, have no metadata. 

Can you confirm that this bug still exists? It might be solved.

**Comment by felixhaeberle** [comment]
@samuelstroschein In which version of the plugin is the fix? v2.2.3? 

`machine translate` still needs to be fixed. 
-> Still applies the messages at the bottom and does not nest them into the correct structure.

**Comment by samuelstroschein** [comment]
@felixhaeberle okay, i'll look into it

**Comment by samuelstroschein** [comment]
@felixhaeberle I will be POOO this week and not investigate this bug. If this is of high importance, please assign someone else.

**Comment by felixhaeberle** [comment]
@NilsJacobsen Can you probably take over? This is important for the growth initiative, because it blocks one PR.

**Comment by NilsJacobsen** [comment]
@felixhaeberle @jonator  is this still happening with the latest version?

**Comment by felixhaeberle** [comment]
@NilsJacobsen No, because osmosis is using the `plugin-json`. Can we backport the behavior to the `plugin-json`?

**Comment by NilsJacobsen** [comment]
@jonator is this still an issue. `jsonPlugin` got several patches?

**Comment by felixhaeberle** [comment]
@NilsJacobsen I tried it once last week, and the issue was still there.

**Comment by NilsJacobsen** [comment]
Okey i feel we should soon remove the deprecated metadata from the jsonPlugin and give it the structure of the i18next plugin. That should fix the error.

**Comment by jonator** [comment]
Will try again with machine-translate and see how it works.

**Comment by samuelstroschein** [comment]
@NilsJacobsen  @yelinz likely also ran into this issue, see https://github.com/inlang/inlang/discussions/1203#discussioncomment-6629759

**Comment by dominikg** [comment]
encountered this issue as well when generating a new locale with `inlang machine translate`
languages/en.json
```json
"foo": {
  "bar":"hello"
}
```

languages/de.json
```json
{
  "foo.bar":"hallo"
}
```

i'd prefer if inlang enforced one style consistently. either no nesting (keys can contain dot, all values must be strings), or nesting (keys must not contain dot, value can be string or object)

This could be a config option on json plugin, read and honored by lint and machine.
Bonus points for extra cli commands to sort by key and convert between flat and nested. Esp. sort can be used in lint-staged, which prevents messy diffs

**Comment by samuelstroschein** [comment]
> edit: without config option you can still detect nesting pretty easy in the ref lang, if there is a value that is object, its nesting, if there is a key that contains dot its flat. both would be mixed and a lint error. 

In theory, we are detecting nesting or not. Seems like a bug. The refactor might solve the bug automatically (release next week)

**Comment by dominikg** [comment]
note i created an entirely new translation file that was empty at first

```shell
echo {} > languages/es.json
inlang machine translate
```
in this case it wouldn't be able to detect nesting from the file being written to.

Adding locales that way can be easily scripted too
create-locales.sh
```shell
#!/bin/sh
for lang in $@;do echo {} > languages/${lang}.json;done
pnpm inlang machine translate
```

or added as a command for the inlang cli?

**Comment by samuelstroschein** [comment]
@dominikg plugins are supposed to handle empty/new language tag files too. Seems like another bug. We should evaluate after the refactor if the bug(s) still exist

PS we can't rely on scripts or the CLI alone. Otherwise, non-devs can't add new languages via the web editor.

**Comment by felixhaeberle** [comment]
@inlang/plugins can you verify that this doesn't occur anymore?

two things discussed in here:
- empty lang file / non-existent lang file
- json children instead of path members for machine translates

**Comment by felixhaeberle** [comment]
Having multiple formattings in multiple files for the same purpose is likely because of human error / buggy applications who I/O the file. So I go for one, unified formatting, detected in the source language file.

**Comment by NiklasBuchfink** [comment]
Related to #1372 https://github.com/inlang/monorepo/pull/1385

**Comment by NiklasBuchfink** [comment]
> * detect once if the message IDs in the source language file are nested or flattened
> * keep the formatting consistent for the whole source language file
> * apply the same formatting for all other language files

I have tested that expected behavior. It now works as proposed.
